pkgs/build-support/setup-hooks/generate-darwin-launcher.sh

```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/env bash
+
+fixupOutputHooks+=('generateDarwinLaunchers $prefix')
+
+# Get a param out of a desktop file. First parameter is the file and the second
+# is a pattern of the key who's value we should fetch.
+getDesktopParam() {
+    local file="$1";
+    local pattern="$2";
+
+    awk -F "=" "/${pattern}/ {print \$2}" "${file}"
+}
+
+# Various icons of different sizes are generated for desktop entries. Here we
+# use the filesize to select the largest icon (which will presumably have the
+# largest resolutioœn) for the launcher.
+getIconFile() {
+    local -r name="$1"
+    local -r candidateIcons=$(find "$out/share/icons/" -name "${name}.png")
+    local biggestSz=0
+    local currentCandidate=""
+
+    for icon in ${candidateIcons}; do
+        local sz=$(stat -c "%s" "${icon}")
+
+        if (( $sz > $biggestSz )); then
+            biggestSz=$sz;
+            currentCandidate="$icon"
+        fi
+    done
+
+    echo "$currentCandidate";
+}
+
+# For a given .desktop file, generate a darwin '.app' launcher for it.
+generateDarwinLauncher() {
+    set -x
+    local -r file="$1"
+    local -r name=$(getDesktopParam "${file}" "^Name")
+    local -r exec=$(getDesktopParam "${file}" "Exec")
+    local -r iconName=$(getDesktopParam "${file}" "Icon")
+    local -r iconFile=$(getIconFile "$iconName")
+
+    mkdir -p "$out/Applications/${name}.app/Contents/MacOS"
+    mkdir -p "$out/Applications/${name}.app/Contents/Resources"
+
+    if [ -n "$iconFile" ]; then
+        png2icns "$out/Applications/${name}.app/Contents/Resources/icon.icns" "$iconFile";
```

Comment:
Can't we include multiple sizes so the OS can select whichever icon fits best? I think `.icns` can contain multiple images. And have you encountered any jpg icons, I think those can be wrapped in `.icns` too?

Reply:
Aha!  Good point.  I've fixed that now.  Although the main compilation is that only *certain* sizses are allowed.  The script will now find all the appropriate icons and covert those.

Reply:
I've revisited the icon code.  Turns out macOS doesn't require `.icns` as I previously thought.  There is also a handy `CFBundleIcons` plist propery which allows you specify an array of icons, rather than having to bundle them all into one `.icns` file.  Much nicer!